8310053: VarHandle and slice handle derived from layout are lacking alignment check

JornVernee opened the pull request:
Add missing alignment checks for the alignment constraint of the root layout of a VarHandle created through `MemoryLayout::varHandle` and a MethodHandle `MemoryLayout::sliceHandle`.

Testing: `jdk_foreign` test suite

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8310068](https://bugs.openjdk.org/browse/JDK-8310068) to be approved

### Issues
 * [JDK-8310053](https://bugs.openjdk.org/browse/JDK-8310053): VarHandle and slice handle derived from layout are lacking alignment check (**Bug** - P3)
 * [JDK-8310068](https://bugs.openjdk.org/browse/JDK-8310068): VarHandle and slice handle derived from layout are lacking alignment check (**CSR**)


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14475/head:pull/14475` \
`$ git checkout pull/14475`

Update a local copy of the PR: \
`$ git checkout pull/14475` \
`$ git pull https://git.openjdk.org/jdk.git pull/14475/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14475`

View PR using the GUI difftool: \
`$ git pr show -t 14475`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14475.diff">https://git.openjdk.org/jdk/pull/14475.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14475#issuecomment-1592049952)

mcimadamore commented on src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:
```diff
@@ -194,16 +197,11 @@ public VarHandle dereferenceHandle(boolean adapt) {
-            // (J, ...) -> J to (J, J, ...) -> J
-            // i.e. new coord is prefixed. Last coord will correspond to innermost layout
-            handle = MethodHandles.collectCoordinates(handle, 1, collector);
+        handle = MethodHandles.collectCoordinates(handle, 1, offsetHandle());
```
nice simplification!

mcimadamore commented on test/jdk/java/foreign/TestLayoutPaths.java:
```diff
@@ -134,6 +135,38 @@ public void testByteOffsetHandleBadRange() {
+
+            String expectedMessage = "Target offset incompatible with alignment constraints: " + struct.byteAlignment();
+
+            try {
```
Can assertThrows be used to simplify this pattern?

JornVernee replied:
Yeah, somewhat. I thought `assertThrows` was only a junit thing, but it looks like TestNG also has an `expectThrows` which does the same and returns the thrown exception so that we can check the message.

mcimadamore commented on src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:
```diff
@@ -380,6 +380,10 @@ default MethodHandle byteOffsetHandle(PathElement... elements) {
      * Additionally, the provided dynamic values must conform to bounds which are derived from the layout path, that is,
      * {@code 0 <= x_i < b_i}, where {@code 1 <= i <= n}, or {@link IndexOutOfBoundsException} is thrown.
      * <p>
+     * The accessed address must be <a href="MemorySegment.html#segment-alignment">aligned</a> according to the {@linkplain
```
Very nice and clear!

JornVernee commented:
I've switched to `expectThrows`. As discussed offline, I've also dropped the alignment check from the inner accessed element, as this is redundant due to layouts enforcing alignment of nested elements already. So, if the root layout is aligned, so is the element. (this turned out to be simper than I initially thought)

mcimadamore reviewed (approved):
Bonus points!

mcimadamore commented:
I'm wondering... for dereference path elements, we already perform a root layout check as part of the memory segment expansion - e.g. we use the target layout to resize the segment (and perform an alignment check). So in that case there would still be duplicate alignment checks.

In such cases, we need to make sure that either (a) only the first var handle in the dereference chain gets the proper alignment adaptation on the input segment or that (b) they all do, but then we omit the alignment checks that are generated as part of calling VH::get on the address layout (but that might be harder to do, as that logic is shared).

JornVernee commented:
> only the first var handle in the dereference chain gets the proper alignment adaptation on the input segment

That would be easy enough to do, we just check for `derefAdapters.length == 0`. I'll add this too.

JornVernee commented:
I've gone ahead and implemented this. I also added a test to make sure that alignment checks still happen when using deref path elements.

I also fixed a minor javadoc issue, in the doc for varHandle. It said that we checked the alignment of the accessed address, but we check the alignment of the base address (a term which is defined earlier).

JornVernee commented:
/integrate

JornVernee commented:
/backport jdk21
